<a href="https://colab.research.google.com/github/Marisagf/ETL_Python/blob/main/ETL_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Projeto 1 - Deletar e Preencher automaticamente os dados ausentes!

In [2]:
# Importanto bibliotecas
import math
import sys, os
import numpy as np
import pandas as pd

In [3]:
# Limitando o nº de colunas
pd.set_option('display.max_columns', 100)

###Importando os Dados

Dataset publico: National Football League (NFL) - 2009 a 2028
Disponível em: https://www.kaggle.com/datasets/maxhorowitz/nflplaybyplay2009to2016

Os arquivos utilizados neste projeto estão disponíveis em:

NFL Play by Play - 2009 - 2017 --> https://www.kaggle.com/code/rtatman/data-cleaning-challenge-handling-missing-values/input?select=NFL+Play+by+Play+2009-2016+%28v3%29.csv

Building_Permits --> https://www.kaggle.com/code/rtatman/data-cleaning-challenge-handling-missing-values/input?select=Building_Permits.csv

In [4]:
# Criamos uma lista para identificar valores ausentes
lista_labels_valores_ausentes = ["n/a", "na", "undefined"]

In [9]:
# Carrega os datasets
# Jogos
dataset = pd.read_csv("/dataset_NFL2009-2017_v4.csv", na_values = lista_labels_valores_ausentes, low_memory=False)
# Licenças de Construção
building_permits = pd.read_csv("/building_permits.csv", na_values = lista_labels_valores_ausentes, low_memory=False)

Atenção: low_memory = False
- Essa opção é necessária, pois a função read_csv() precisa ler o arquivo inteiro para só então determinar o tipo de dado de cada coluna. Como o arquivo de dados da NFL é muito grande, ocorre um erro, se não o fizermos dessa forma. Isso faz com que o Pandas não determine automaticamente os tipos de dados. Você pode ver que isso é verdade, ao buscar as informações do datasetm como mostrado no próximo comando:

In [10]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 407688 entries, 0 to 407687
Columns: 102 entries, Date to Season
dtypes: float64(33), int64(31), object(38)
memory usage: 317.3+ MB


In [11]:
dataset.sample(6) # Informa a quantidade de linhas e colunas.

Date      GameID  Drive  qtr  down   time  TimeUnder  TimeSecs  \
290333  2015-11-01  2015110108     21    4   3.0  13:33         14     813.0   
353059  2016-12-11  2016121110     26    5   2.0  12:07         13    -173.0   
402893  2017-12-24  2017122404     11    3   3.0  05:48          6    1248.0   
140604  2012-09-23  2012092307      6    1   2.0  01:58          2    2818.0   
397352  2017-12-10  2017121004     11    2   2.0  05:11          6    2111.0   
297014  2015-11-16  2015111600      9    2   2.0  03:10          4    1990.0   

        PlayTimeDiff SideofField  yrdln  yrdline100  ydstogo  ydsnet  \
290333          42.0         ARI   26.0        74.0        4      18   
353059          44.0         NYJ   41.0        59.0        6      12   
402893          39.0         BUF   10.0        10.0        5      59   
140604           0.0         BUF   41.0        59.0        6       4   
397352          26.0          SF   34.0        66.0        8      31   
297014          38.0         CIN   26.0        74.0        9       4   

        GoalToGo  FirstDown posteam DefensiveTeam  \
290333       0.0        1.0     ARI           CLE   
353059       0.0        1.0     NYJ            SF   
402893       0.0        0.0      NE           BUF   
140604       0.0        0.0     BUF           CLE   
397352       0.0        1.0      SF           HOU   
297014       0.0        0.0     CIN           HOU   

                                                     desc  PlayAttempted  \
290333  (13:33) (Shotgun) C.Palmer pass short left to ...              1   
353059  (12:07) (Shotgun) B.Powell left end to NYJ 49 ...              1   
402893  (5:48) (Shotgun) T.Brady sacked at BUF 16 for ...              1   
140604  (1:58) (Shotgun) R.Fitzpatrick pass incomplete...              1   
397352  (5:11) J.Garoppolo pass deep left to K.Juszczy...              1   
297014  (3:10) (Shotgun) A.Dalton pass short middle to...              1   

        Yards.Gained  sp  Touchdown ExPointResult TwoPointConv DefTwoPoint  \
290333            12   0          0           NaN          NaN         NaN   
353059             8   0          0           NaN          NaN         NaN   
402893            -6   0          0           NaN          NaN         NaN   
140604             0   0          0           NaN          NaN         NaN   
397352            29   0          0           NaN          NaN         NaN   
297014             3   0          0           NaN          NaN         NaN   

        Safety  Onsidekick PuntResult PlayType         Passer   Passer_ID  \
290333       0           0        NaN     Pass       C.Palmer  00-0021429   
353059       0           0        NaN      Run            NaN         NaN   
402893       0           0        NaN     Sack            NaN  00-0019596   
140604       0           0        NaN     Pass  R.Fitzpatrick  00-0023682   
397352       0           0        NaN     Pass    J.Garoppolo  00-0031345   
297014       0           0        NaN     Pass       A.Dalton  00-0027973   

        PassAttempt      PassOutcome PassLength  AirYards  YardsAfterCatch  \
290333            1         Complete      Short         4                8   
353059            0              NaN        NaN         0                0   
402893            0              NaN        NaN         0                0   
140604            1  Incomplete Pass       Deep        17                0   
397352            1         Complete       Deep        26                3   
297014            1         Complete      Short         0                3   

        QBHit PassLocation  InterceptionThrown Interceptor    Rusher  \
290333      1         left                   0         NaN       NaN   
353059      0          NaN                   0         NaN  B.Powell   
402893      1          NaN                   0         NaN       NaN   
140604      0       middle                   0         NaN       NaN   
397352      1         left                   0   

In [12]:
dataset.shape

(407688, 102)

In [13]:
building_permits.sample(6) # Os dados são exibidos em formato tabular com algumas colunas com valores NaN. Esses valores são chamados de valores ausentes (Not a Number).

Permit Number  Permit Type  Permit Type Definition  \
127510  201606160101            8  otc alterations permit   
16106   201306189859            8  otc alterations permit   
18985   201307182170            8  otc alterations permit   
156440       M783867            8  otc alterations permit   
30727   201311121601            8  otc alterations permit   
143153  201611213212            8  otc alterations permit   

       Permit Creation Date Block   Lot  Street Number Street Number Suffix  \
127510           06/16/2016  0316  018A            445                  NaN   
16106            06/18/2013  1672   032            820                  NaN   
18985            07/18/2013  0174   008            710                  NaN   
156440           05/09/2017  2654   047            300                  NaN   
30727            11/12/2013  3514   030           1663                  NaN   
143153           11/21/2016  5338   042           1747                  NaN   

       Street Name Street Suffix  Unit Unit Suffix  \
127510       Geary            St   0.0         NaN   
16106         30th            Av   NaN         NaN   
18985      Sansome            St   NaN         NaN   
156440     Corbett            Av   0.0         NaN   
30727      Mission            St   NaN         NaN   
143153     Quesada            Av   NaN         NaN   

                                              Description Current Status  \
127510  addition to permit 201503060168 to show all ro...       complete   
16106   repair existing stairs in kind for dry rot wat...         issued   
18985   basement/1st/2nd/3rd floors-t.i to include new...         issued   
156440                                       street space         issued   
30727   7th floor: add 4 sprinkler heads. move one spr...       complete   
143153                                          reroofing       complete   

       Current Status Date  Filed Date Issued Date Completed Date  \
127510          01/24/2017  06/16/2016  06/20/2016     01/24/2017   
16106           06/20/2013  06/18/2013  06/20/2013            NaN   
18985           08/01/2013  07/18/2013  08/01/2013            NaN   
156440          05/09/2017  05/09/2017  05/09/2017            NaN   
30727           12/10/2013  11/12/2013  11/12/2013     12/10/2013   
143153          12/01/2016  11/21/2016  11/21/2016     12/01/2016   

       First Construction Document Date Structural Notification  \
127510                       06/20/2016                     NaN   
16106                        06/20/2013                     NaN   
18985                        08/01/2013                     NaN   
156440                       05/09/2017                     NaN   
30727                        11/12/2013                     NaN   
143153                       11/21/2016                     NaN   

        Number of Existing Stories  Number of Proposed Stories  \
127510                         3.0                         3.0   
16106                          3.0                         3.0   
18985                          3.0                         3.0   
156440                         NaN                         NaN   
30727                          7.0                         7.0   
143153                         2.0                         2.0   

       Voluntary Soft-Story Retrofit Fire Only Permit Permit Expiration Date  \
127510                           NaN              NaN             06/15/2017   
16106                            NaN              NaN             06/15/2014   
18985                            NaN              NaN             07/16/2016   
156440                           NaN              NaN                    NaN   
30727                            NaN                Y             11/07/2014   
143153                           NaN              NaN             11/16/2017   

        Estimated Cost  Revised Cost       Existing Use  Existing Units  \
127510             1.0           1.0            theater     

Depois de ler os dados, descobrimos que ambos os conjuntos de dados têm valores ausentes.

Nosso próximo passo será:
-  calcular o número de valores ausentes que temos em cada coluna.

Para contar os valores nulos, o Pandas têm a função isnull(). Como o dataset principal possui 102 colunas, analisaremos as primeiras dez colunas contendo valores ausentes.

In [14]:
contagem_valores_ausentes = dataset.isnull().sum()
contagem_valores_ausentes[0:20]

Date                 0
GameID               0
Drive                0
qtr                  0
down             61154
time               224
TimeUnder            0
TimeSecs           224
PlayTimeDiff       444
SideofField        528
yrdln              840
yrdline100         840
ydstogo              0
ydsnet               0
GoalToGo           840
FirstDown        28811
posteam          24992
DefensiveTeam    24992
desc                 2
PlayAttempted        0
dtype: int64

Cada nome de coluna e o número associado indicam o número de valores ausentes!

Vamos verificar a porcentagem dos valores ausentes em nosso conjunto de dados.

Para este cálculo de porcentagem, usaremos a combinação das funções prod() do Numpy e shape do Pandas.

In [16]:
# Calculando os totais de celulas e de celulas com valores ausentes (total_nulos)
total_celulas = np.prod(dataset.shape)
total_nulos = contagem_valores_ausentes.sum()
# Calculando o percentual de dados ausentes
print(f"Percentual de Dados ausentes: {(total_nulos/total_celulas):.2%}")

Percentual de Dados ausentes: 27.67%


Agora é sua vez de aplicar as mesmas etapas no conjunto de dados building_permits e verificar a porcentagem de valores ausentes.


In [17]:

contagem_valores_ausentes2 = building_permits.isnull().sum()
contagem_valores_ausentes2[0:20]
contagem_valores_ausentes2 = building_permits.isnull().sum()

In [18]:
# Calculando os totais de celulas e de celulas com valores ausentes (total_nulos)
total_celulas2 = np.prod(building_permits.shape)
total_nulos2 = contagem_valores_ausentes2.sum()
# Calculando o percentual de dados ausentes
print(f"Percentual de Dados ausentes: {(total_nulos2/total_celulas2):.2%}")

Percentual de Dados ausentes: 26.26%


#Eliminar (Drop) valores ausentes.

 - Observação:

Geralmente essa abordagem não é recomendada para projetos importantes! Vale a pena dedicar um tempo para examinar seus dados e realmente examinar todas as colunas com valores ausentes, uma a uma, para realmente conhecer seu conjunto de dados

Se você tem certeza de que deseja descartar linhas com valores ausentes, o pandas tem uma função útil, dropna() para ajudá-lo a fazer isso. Vamos experimentá-lo em nosso conjunto de dados da NFL!

- Atenção:


Se utilizar o comando: dataset.dropna() ele removerá todas as linhas que contém valores ausentes. Neste caso especificamente, resultará em um dataset vazio, pois todas a linhas possuem ao menos um valor ausente. Assim, é mais interessante, indicarmos que sejam apagados apenas as colunas que possuem dados ausentes. Isso é possível, passando um parâmetro para esta função, indicando o eixo: no caso axis="columns".



In [20]:
dataset_semNA = dataset.dropna(axis='columns')
dataset_semNA.head()


(407688, 37)

In [21]:
# Fazendo a comparação entre os datasets
print(f"Colunas no dataset original {dataset.shape[1]}")
print(f"Colunas no dataset sem NA   {dataset_semNA.shape[1]}")

Colunas no dataset original 102
Colunas no dataset sem NA   37


Com essa aplicação perdemos muitas informações do nosso dataset original. Todavia, temos um dataset agora sem valores ausentes!!

#Preenchendo valores ausentes automaticamente
Outra opção é tentar preencher os valores que faltam. Para fins didáticos, estou recortando apenas uma parte do dataset original, para que nosso processo fique mais elucidativo.

In [22]:
# Obtendo uma pequena parte do dataset original
subconjuntoNFL = dataset.loc[:, "EPA":"Season"].head()
subconjuntoNFL

EPA    airEPA    yacEPA  Home_WP_pre  Away_WP_pre  Home_WP_post  \
0  2.014474       NaN       NaN     0.485675     0.514325      0.546433   
1  0.077907 -1.068169  1.146076     0.546433     0.453567      0.551088   
2 -1.402760       NaN       NaN     0.551088     0.448912      0.510793   
3 -1.712583  3.318841 -5.031425     0.510793     0.489207      0.461217   
4  2.097796       NaN       NaN     0.461217     0.538783      0.558929   

   Away_WP_post  Win_Prob       WPA    airWPA    yacWPA  Season  
0      0.453567  0.485675  0.060758       NaN       NaN    2009  
1      0.448912  0.546433  0.004655 -0.032244  0.036899    2009  
2      0.489207  0.551088 -0.040295       NaN       NaN    2009  
3      0.538783  0.510793 -0.049576  0.106663 -0.156239    2009  
4      0.441071  0.461217  0.097712       NaN       NaN    2009

Podemos usar a função fillna() do Panda para preencher os valores ausentes em um dataframe para nós. Uma opção que temos é especificar o que queremos que os valores NaN sejam substituídos. Vamos fazer a substituição de todos os valores NaN por 0 (zero).

In [23]:
# substituindo todos os valores NaN por 0
subconjuntoNFL.fillna(0)

EPA    airEPA    yacEPA  Home_WP_pre  Away_WP_pre  Home_WP_post  \
0  2.014474  0.000000  0.000000     0.485675     0.514325      0.546433   
1  0.077907 -1.068169  1.146076     0.546433     0.453567      0.551088   
2 -1.402760  0.000000  0.000000     0.551088     0.448912      0.510793   
3 -1.712583  3.318841 -5.031425     0.510793     0.489207      0.461217   
4  2.097796  0.000000  0.000000     0.461217     0.538783      0.558929   

   Away_WP_post  Win_Prob       WPA    airWPA    yacWPA  Season  
0      0.453567  0.485675  0.060758  0.000000  0.000000    2009  
1      0.448912  0.546433  0.004655 -0.032244  0.036899    2009  
2      0.489207  0.551088 -0.040295  0.000000  0.000000    2009  
3      0.538783  0.510793 -0.049576  0.106663 -0.156239    2009  
4      0.441071  0.461217  0.097712  0.000000  0.000000    2009

In [24]:
# Preenchendo os valores ausentes com os valores que vem logo a seguir dos NaN

# bfill(axis=0): Este método preenche valores ausentes utilizando o preenchimento para trás (backward fill).
# Isso significa que, para cada célula que tem um valor ausente (NaN), ela será preenchida com o valor da célula
# imediatamente posterior (considerando o eixo 0, ou seja, as linhas).
# Se uma célula não tiver um valor posterior, ela permanecerá como NaN.
# fillna(0): Após a aplicação do bfill, o método fillna é chamado para preencher todos os valores ausentes
# restantes com zero (0). Assim, se ainda houver células com valores ausentes após o preenchimento para trás,
# essas células serão substituídas pelo valor zero.

subconjuntoNFL.bfill(axis=0).fillna(0)

EPA    airEPA    yacEPA  Home_WP_pre  Away_WP_pre  Home_WP_post  \
0  2.014474 -1.068169  1.146076     0.485675     0.514325      0.546433   
1  0.077907 -1.068169  1.146076     0.546433     0.453567      0.551088   
2 -1.402760  3.318841 -5.031425     0.551088     0.448912      0.510793   
3 -1.712583  3.318841 -5.031425     0.510793     0.489207      0.461217   
4  2.097796  0.000000  0.000000     0.461217     0.538783      0.558929   

   Away_WP_post  Win_Prob       WPA    airWPA    yacWPA  Season  
0      0.453567  0.485675  0.060758 -0.032244  0.036899    2009  
1      0.448912  0.546433  0.004655 -0.032244  0.036899    2009  
2      0.489207  0.551088 -0.040295  0.106663 -0.156239    2009  
3      0.538783  0.510793 -0.049576  0.106663 -0.156239    2009  
4      0.441071  0.461217  0.097712  0.000000  0.000000    2009

In [26]:
# Salvando os dados
subconjuntoNFL.to_csv('/dataset_NFL_limpo1.csv')